fdbserver/KeyValueStoreShardedRocksDB.actor.cpp

```diff
@@ -148,6 +150,200 @@ class RocksDBEventListener : public rocksdb::EventListener {
 	UID logId;
 };
 
+// Returns string representation of RocksDB flush reason.
+// Flush reason code:
+// https://github.com/facebook/rocksdb/blob/63a5125a5220d953bf504daf33694f038403cc7c/include/rocksdb/listener.h#L164-L181
+// This function needs to be updated when flush code changes.
+std::string getFlushReason(const FlushReason& reason) {
+	switch (reason) {
+	case FlushReason::kOthers:
+		return format("%d Others", reason);
+	case FlushReason::kGetLiveFiles:
+		return format("%d GetLiveFiles", reason);
+	case FlushReason::kShutDown:
+		return format("%d ShutDown", reason);
+	case FlushReason::kExternalFileIngestion:
+		return format("%d ExternalFileIngestion", reason);
+	case FlushReason::kManualCompaction:
+		return format("%d ManualCompaction", reason);
+	case FlushReason::kWriteBufferManager:
+		return format("%d WriteBufferManager", reason);
+	case FlushReason::kWriteBufferFull:
+		return format("%d WriteBufferFull", reason);
+	case FlushReason::kTest:
+		return format("%d Test", reason);
+	case FlushReason::kDeleteFiles:
+		return format("%d DeleteFiles", reason);
+	case FlushReason::kAutoCompaction:
+		return format("%d AutoCompaction", reason);
+	case FlushReason::kManualFlush:
+		return format("%d ManualFlush", reason);
+	case FlushReason::kErrorRecovery:
+		return format("%d ErrorRecovery", reason);
+	case FlushReason::kErrorRecoveryRetryFlush:
+		return format("%d ErrorRecoveryRetryFlush", reason);
+	case FlushReason::kWalFull:
+		return format("%d WalFull", reason);
+	default:
+		return format("%d Unknown", reason);
+	}
+}
+
+struct FlushStats {
+	std::string cfName;
+	uint64_t threadId;
+	FlushReason reason;
+	double whenStart;
+	FlushStats(std::string cfName, uint64_t threadId, FlushReason reason, double whenStart)
+	  : cfName(cfName), threadId(threadId), reason(reason), whenStart(whenStart) {}
+};
+
+class RocksDBFlushListener : public rocksdb::EventListener {
+public:
+	RocksDBFlushListener(UID id) : logId(id) {}
+	void OnFlushBegin(rocksdb::DB* db, const rocksdb::FlushJobInfo& info) override {
+		std::lock_guard<std::mutex> lock(mutex);
```

Comment:
Use mutex here may cause deadlock.

Reply:
Is SpinLock (implemented by atomic operation) acceptable?